packages/libp2p/src/dcutr/dcutr.ts

```diff
@@ -153,57 +202,65 @@ export class DefaultDCUtRService implements Startable {
         // to measure RTT of the relay connection.
-        log('B sending connect')
+        logB('sending CONNECT to A')
         const connectTimer = Date.now()
         await pb.write({
           type: HolePunch.Type.CONNECT,
           observedAddresses: this.addressManager.getAddresses().map(ma => ma.bytes)
         }, options)
 
-        log('B receiving connect')
+        logB('waiting for CONNECT from A')
         // 4. Upon receiving the Connect, B sends a Sync message
         const connect = await pb.read(options)
+        logB('received CONNECT from A')
 
         if (connect.type !== HolePunch.Type.CONNECT) {
-          log('A sent wrong message type')
+          logB('A sent the wrong message type')
           throw new CodeError('DCUtR message type was incorrect', codes.ERR_INVALID_MESSAGE)
         }
 
         const multiaddrs = this.getDialableMultiaddrs(connect.observedAddresses)
 
         if (multiaddrs.length === 0) {
-          log('A did not have any dialable multiaddrs')
+          logB('A did not have any dialable multiaddrs')
           throw new CodeError('DCUtR connect message had no multiaddrs', codes.ERR_INVALID_MESSAGE)
         }
 
         const rtt = Date.now() - connectTimer
 
-        log('A sending sync, rtt %dms', rtt)
+        logB('sending SYNC to A, rtt %dms', rtt)
         await pb.write({
           type: HolePunch.Type.SYNC,
           observedAddresses: []
         }, options)
 
-        log('A waiting for half RTT')
+        logB('waiting for half RTT')
         // ..and starts a timer for half the RTT measured from the time between
         // sending the initial Connect and receiving the response
+        // TODO: Find a way to simulate network latency for tests, because rtt is always <= 4ms
         await delay(rtt / 2)
 
         // TODO: when we have a QUIC transport, the dial step is different - for
         // now we only have tcp support
         // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
 
-        log('B dialing', multiaddrs)
-        // Upon expiry of the timer, B dials the address to A.
-        await this.connectionManager.openConnection(multiaddrs, {
+        logB('dialing', multiaddrs)
+        // Upon expiry of the timer (await delay above), B dials the address to A.
+        const newDirectConnection = await this.connectionManager.openConnection(multiaddrs, {
           signal: options.signal,
           priority: DCUTR_DIAL_PRIORITY
         })
+        logB('dialing succeeded. connection ID =', newDirectConnection.id)
+
+        logB('DCUtR to %p succeeded, A SHOULD close relayed connection', relayedConnection.remotePeer)
 
-        log('DCUtR to %p succeeded, closing relayed connection', relayedConnection.remotePeer)
+        // Once a single connection has been established, A SHOULD cancel all outstanding connection attempts.
         await relayedConnection.close(options)
```

Comment:
at the very least, finding some way to simulate network latency.. or even simulate being behind a NAT (instead of the test running entirely on the same network) would be great.

Reply:
> Has there been any discussion into supported eventEmitters on Connections? it would certainly help here

I'm not sure what you're asking for here - could you elaborate?

> All new streams should be opened in the direct connection, while the relay connection should be closed after a grace period.

This is really what we need the [stream migration protocol](https://github.com/libp2p/specs/issues/328) for.

That said, right now (as of `0.46.x`) a relayed connection is marked as "transient" and all streams have to opt-in to running on transient streams (bitswap, for example, will not run on a transient stream).

The only protocols we have that opt-in by default are circuit relay, identify, ping and now DCUtR.

Circuit relay has already run (since we have a relayed connection) as has identify (since we were notified by the registrar) so closing the relayed connection might disrupt ping but 🤷 

Stream migration would handle when the user has a custom protocol that has opted in to running on transient streams but (famous last words) that feels like an uncommon use case.

> finding some way to simulate network latency.. or even simulate being behind a NAT

Yeah, that's why the test suite for the connection upgrade has `// TODO: how to test this?` 😄

Reply:
> > Has there been any discussion into supported eventEmitters on Connections? it would certainly help here

> I'm not sure what you're asking for here - could you elaborate?

In the following scenario: 

* A has an outgoing Connection to B, id: 'foo'
* B has an incoming Connection from A, id: 'bar' // I presume the id's are different and that we don't transmit this internal representation of a connection
* These connections are representative of the same connection at the networking layer // maybe this isn't the case?

Does A have any awareness, through Connection='foo', that B calls `close` on Connection='bar' ? How is that propagated? 

I was expecting that I would be able to listen for `Connection.on('remoteClose')` or similar, so if the other side of a connection closes it, I can handle operations that may depend on the assumption that the connection is still viable. 

Does that make more sense?

---

> This is really what we need the https://github.com/libp2p/specs/issues/328 for.

Thanks for the link. I will dive into this 

---

> Yeah, that's why the test suite for the connection upgrade has // TODO: how to test this? 😄

yea but I had to give it a shot! Also, at least now we know the logic works, even if we're not validating that the actual direct connection wasn't possible before. i.e. the test now confirms that DCUtR flow is operating as expected, but not that DCUtR is actually solving a via HolePunching. 

And now it's more specific!

I was wondering if there was some way we could mimic a NAT via connectionGater or block/deny lists, but I couldn't think of a way to make it work without blocking the simultaneous connect dial attempt.
---

Reply:
> I presume the id's are different and that we don't transmit this internal representation of a connection

That's correct, A and B have no knowledge of each other's connection IDs or even if the implementation at the other end of the connection even has connection IDs.

> Does A have any awareness, through Connection='foo', that B calls close on Connection='bar' ? How is that propagated?

No, the underlying transport will notice that the socket is closed, and close the `MultiaddrConnection` that underlies the Connection.  Any attempt to open new streams or send data etc will then throw.

Basically you assume the connection is viable, if it's not, whatever you are trying to do with it will throw.